### PR TITLE
RFE: tests: add test for system time change auditing

### PR DESCRIPTION
**Pull request by WOnder93**
Wiki: https://github.com/linux-audit/audit-kernel/wiki/RFE-More-detailed-auditing-of-changes-to-system-clock
Kernel issue: https://github.com/linux-audit/audit-kernel/issues/10

Signed-off-by: Ondrej Mosnacek <omosnace@redhat.com>

### CODE REVIEW

**Comment by pcmoore** [comment]
This test appears to be badly broken on my system:

```
Running as   user    root
        with context unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
        on   system  Fedora

exec_execve/test ......... ok   
exec_name/test ........... ok     
file_create/test ......... ok   
file_delete/test ......... ok   
file_rename/test ......... ok   
filter_exclude/test ...... ok     
filter_sessionid/test .... ok   
login_tty/test ........... ok   
lost_reset/test .......... ok   
netfilter_pkt/test ....... ok     
syscalls_file/test ....... ok   
syscall_module/test ...... ok   
syscall_socketcall/test .. ok   
time_change/test ......... 1/68 # Test 2 got: "256" (time_change/test at line 72)
#   Expected: "0"
#  time_change/test line 72 is:     ok( $result, 0 );
# Failed test 3 in time_change/test at line 87
#  time_change/test line 87 is:     ok($found_injoffset);
# Test 4 got: "256" (time_change/test at line 91)
#   Expected: "0"
#  time_change/test line 91 is:     ok( $result, 0 );
# Test 6 got: "256" (time_change/test at line 72 fail #2)
#   Expected: "0"
# Failed test 7 in time_change/test at line 87 fail #2
# Test 8 got: "256" (time_change/test at line 91 fail #2)
#   Expected: "0"
# Failed test 10 in time_change/test at line 141
#  time_change/test line 141 is:         ok($found_injoffset);
# Test 11 got: "256" (time_change/test at line 146)
#    Expected: "0"
#  time_change/test line 146 is:         ok( $result, 0 );
# Failed test 13 in time_change/test at line 141 fail #2
# Test 14 got: "256" (time_change/test at line 146 fail #2)
#    Expected: "0"
# Test 22 got: "0" (time_change/test at line 193)
#    Expected: "2"
#  time_change/test line 193 is:     ok( $found_adjntpval_tot, 2 );
# Failed test 23 in time_change/test at line 196
#  time_change/test line 196 is:     ok($found_adjntpval);
# Test 24 got: "256" (time_change/test at line 200)
#    Expected: "0"
#  time_change/test line 200 is:     ok( $result, 0 );
# Test 26 got: "0" (time_change/test at line 193 fail #2)
#    Expected: "2"
# Failed test 27 in time_change/test at line 196 fail #2
# Test 28 got: "256" (time_change/test at line 200 fail #2)
#    Expected: "0"
# Test 30 got: "0" (time_change/test at line 193 fail #3)
#    Expected: "2"
# Failed test 31 in time_change/test at line 196 fail #3
# Test 32 got: "256" (time_change/test at line 200 fail #3)
#    Expected: "0"
time_change/test ......... 33/68 # Test 34 got: "0" (time_change/test at line 193 fail #4)
#    Expected: "2"
# Failed test 35 in time_change/test at line 196 fail #4
# Test 36 got: "256" (time_change/test at line 200 fail #4)
#    Expected: "0"
# Test 38 got: "0" (time_change/test at line 193 fail #5)
#    Expected: "2"
# Failed test 39 in time_change/test at line 196 fail #5
# Test 40 got: "256" (time_change/test at line 200 fail #5)
#    Expected: "0"
# Test 42 got: "0" (time_change/test at line 193 fail #6)
#    Expected: "2"
# Failed test 43 in time_change/test at line 196 fail #6
# Test 44 got: "256" (time_change/test at line 200 fail #6)
#    Expected: "0"
# Test 46 got: "0" (time_change/test at line 193 fail #7)
#    Expected: "2"
# Failed test 47 in time_change/test at line 196 fail #7
# Test 48 got: "256" (time_change/test at line 200 fail #7)
#    Expected: "0"
# Test 50 got: "0" (time_change/test at line 193 fail #8)
#    Expected: "2"
# Failed test 51 in time_change/test at line 196 fail #8
# Test 52 got: "256" (time_change/test at line 200 fail #8)
#    Expected: "0"
# Test 54 got: "0" (time_change/test at line 193 fail #9)
#    Expected: "2"
# Failed test 55 in time_change/test at line 196 fail #9
# Test 56 got: "256" (time_change/test at line 200 fail #9)
#    Expected: "0"
# Test 58 got: "0" (time_change/test at line 193 fail #10)
#    Expected: "2"
# Failed test 59 in time_change/test at line 196 fail #10
# Test 60 got: "256" (time_change/test at line 200 fail #10)
#    Expected: "0"
# Test 62 got: "0" (time_change/test at line 193 fail #11)
#    Expected: "2"
# Failed test 63 in time_change/test at line 196 fail #11
# Test 64 got: "256" (time_change/test at line 200 fail #11)
#    Expected: "0"
# Test 66 got: "0" (time_change/test at line 193 fail #12)
#    Expected: "2"
# Failed test 67 in time_change/test at line 196 fail #12
# Test 68 got: "256" (time_change/test at line 200 fail #12)
#    Expected: "0"
time_change/test ......... Failed 46/68 subtests 
user_msg/test ............ ok   

Test Summary Report
-------------------
time_change/test       (Wstat: 0 Tests: 68 Failed: 46)
  Failed tests:  2-4, 6-8, 10-11, 13-14, 22-24, 26-28, 30-32
                34-36, 38-40, 42-44, 46-48, 50-52, 54-56
                58-60, 62-64, 66-68
Files=15, Tests=172, 17 wallclock secs ( 0.08 usr  0.04 sys +  4.90 cusr  2.26 csys =  7.28 CPU)
Result: FAIL
Failed 1/15 test programs. 46/172 subtests failed.
```

Some additional information on my test system:

```
# uname -a
Linux cisco-rawhide-1 5.2.0-0.rc0.git1.1.1.secnext.fc31.x86_64 #1 SMP Tue May 7 17:24:41 EDT 2019 x86_64 x86_64 x86_64 GNU/Linux
# rpm -q audit
audit-3.0-0.7.20190326git03e7489.fc31.x86_64
# git log --oneline | head
fea8985 tests: add test for system time change auditing
da8b17f tests/Makefile: do not override the PATH variable
38f7a5f exec_name: remove funky characters from the last commit
57d0261 tests: test also exclude filter by executable name
6d65bf1 tests: update rules to select correct arch
7ff3e43 Revert "all: additional code formatting fixes"
13f2287 all: additional code formatting fixes
fa15461 lost_reset: fix code formatting
5a6f30a lost_reset: spin until msg lost
c0aa026 tests: reset_lost: tune up to avoid false positives
```

**Comment by pcmoore** [comment]
Some additional output:

```
# cd tests/time_change
# ./test
1..68
# Running under perl version 5.028002 for linux
# Current time local: Wed May  8 15:06:16 2019
# Current time GMT:   Wed May  8 19:06:16 2019
# Using Test.pm version 1.31
arch elf mapping not found
ok 1
not ok 2
# Test 2 got: "256" (./test at line 72)
#   Expected: "0"
#  ./test line 72 is:     ok( $result, 0 );
not ok 3
# Failed test 3 in ./test at line 87
#  ./test line 87 is:     ok($found_injoffset);
not ok 4
# Test 4 got: "256" (./test at line 91)
#   Expected: "0"
#  ./test line 91 is:     ok( $result, 0 );
arch elf mapping not found
ok 5
not ok 6
# Test 6 got: "256" (./test at line 72 fail #2)
#   Expected: "0"
not ok 7
# Failed test 7 in ./test at line 87 fail #2
not ok 8
# Test 8 got: "256" (./test at line 91 fail #2)
#   Expected: "0"
arch elf mapping not found
ok 9
not ok 10
# Failed test 10 in ./test at line 141
#  ./test line 141 is:         ok($found_injoffset);
not ok 11
# Test 11 got: "256" (./test at line 146)
#    Expected: "0"
#  ./test line 146 is:         ok( $result, 0 );
arch elf mapping not found
ok 12
not ok 13
# Failed test 13 in ./test at line 141 fail #2
not ok 14
# Test 14 got: "256" (./test at line 146 fail #2)
#    Expected: "0"
arch elf mapping not found
ok 15
ok 16
not ok 17
# Test 17 got: "256" (./test at line 138)
#    Expected: "0"
#  ./test line 138 is:         ok( $result, 0 );
arch elf mapping not found
ok 18
ok 19
not ok 20
# Test 20 got: "256" (./test at line 138 fail #2)
#    Expected: "0"
arch elf mapping not found
ok 21
not ok 22
# Test 22 got: "0" (./test at line 193)
#    Expected: "2"
#  ./test line 193 is:     ok( $found_adjntpval_tot, 2 );
not ok 23
# Failed test 23 in ./test at line 196
#  ./test line 196 is:     ok($found_adjntpval);
not ok 24
# Test 24 got: "256" (./test at line 200)
#    Expected: "0"
#  ./test line 200 is:     ok( $result, 0 );
arch elf mapping not found
ok 25
not ok 26
# Test 26 got: "0" (./test at line 193 fail #2)
#    Expected: "2"
not ok 27
# Failed test 27 in ./test at line 196 fail #2
not ok 28
# Test 28 got: "256" (./test at line 200 fail #2)
#    Expected: "0"
arch elf mapping not found
ok 29
not ok 30
# Test 30 got: "0" (./test at line 193 fail #3)
#    Expected: "2"
not ok 31
# Failed test 31 in ./test at line 196 fail #3
not ok 32
# Test 32 got: "256" (./test at line 200 fail #3)
#    Expected: "0"
arch elf mapping not found
ok 33
not ok 34
# Test 34 got: "0" (./test at line 193 fail #4)
#    Expected: "2"
not ok 35
# Failed test 35 in ./test at line 196 fail #4
not ok 36
# Test 36 got: "256" (./test at line 200 fail #4)
#    Expected: "0"
arch elf mapping not found
ok 37
not ok 38
# Test 38 got: "0" (./test at line 193 fail #5)
#    Expected: "2"
not ok 39
# Failed test 39 in ./test at line 196 fail #5
not ok 40
# Test 40 got: "256" (./test at line 200 fail #5)
#    Expected: "0"
arch elf mapping not found
ok 41
not ok 42
# Test 42 got: "0" (./test at line 193 fail #6)
#    Expected: "2"
not ok 43
# Failed test 43 in ./test at line 196 fail #6
not ok 44
# Test 44 got: "256" (./test at line 200 fail #6)
#    Expected: "0"
arch elf mapping not found
ok 45
not ok 46
# Test 46 got: "0" (./test at line 193 fail #7)
#    Expected: "2"
not ok 47
# Failed test 47 in ./test at line 196 fail #7
not ok 48
# Test 48 got: "256" (./test at line 200 fail #7)
#    Expected: "0"
arch elf mapping not found
ok 49
not ok 50
# Test 50 got: "0" (./test at line 193 fail #8)
#    Expected: "2"
not ok 51
# Failed test 51 in ./test at line 196 fail #8
not ok 52
# Test 52 got: "256" (./test at line 200 fail #8)
#    Expected: "0"
arch elf mapping not found
ok 53
not ok 54
# Test 54 got: "0" (./test at line 193 fail #9)
#    Expected: "2"
not ok 55
# Failed test 55 in ./test at line 196 fail #9
not ok 56
# Test 56 got: "256" (./test at line 200 fail #9)
#    Expected: "0"
arch elf mapping not found
ok 57
not ok 58
# Test 58 got: "0" (./test at line 193 fail #10)
#    Expected: "2"
not ok 59
# Failed test 59 in ./test at line 196 fail #10
not ok 60
# Test 60 got: "256" (./test at line 200 fail #10)
#    Expected: "0"
arch elf mapping not found
ok 61
not ok 62
# Test 62 got: "0" (./test at line 193 fail #11)
#    Expected: "2"
not ok 63
# Failed test 63 in ./test at line 196 fail #11
not ok 64
# Test 64 got: "256" (./test at line 200 fail #11)
#    Expected: "0"
arch elf mapping not found
ok 65
not ok 66
# Test 66 got: "0" (./test at line 193 fail #12)
#    Expected: "2"
not ok 67
# Failed test 67 in ./test at line 196 fail #12
not ok 68
# Test 68 got: "256" (./test at line 200 fail #12)
#    Expected: "0"
```

**Comment by WOnder93** [comment]
I think this is because you don't have a recent enough userspace that would recognize the new record types by name.

> `audit-3.0-0.7.20190326git03e7489.fc31.x86_64`
> `-------------------------^^^^^^^`

linux-audit/audit-userspace@03e7489 is a few commits behind linux-audit/audit-userspace@e2fa559b.

**Comment by pcmoore** [comment]
Bummer, I thought that was released into Rawhide already.  Oh well.

Thanks for the quick reply, please update/ping this issue when Steve has update his audit userspace packages.

**Comment by WOnder93** [comment]
(force-pushed to add sync code as in f7023d428cc5 + rebase on master)

**Comment by pcmoore** [comment]
Just tried this again and saw similar problems.  @stevegrubb when do you expect to push an updated audit userspace to Rawhide?

**Comment by stevegrubb** [comment]
rawhide was from the Nov 4 tree. I have no plans to update rawhide or Fedora as of right now because nothing important has landed.

**Comment by pcmoore** [comment]
@WOnder93 does this work for you on Fedora Rawhide or a normal stable Fedora release?

**Comment by WOnder93** [comment]
It fails for me as well on an old Rawhide kernel 5.4.0-0.rc0.git3.1.fc32.x86_64 (the current one has networking issues), but passes on F31 kernel... It seems the newer kernel has some trouble tagging the syscall record with the right key based on the rule. Perhaps some syscall numbering bug somewhere?

**Comment by pcmoore** [comment]
@WOnder93 this is still failing for me with Linux v5.6.0-rcX; please take a closer look into this and either fix the test of fix the kernel (depending on what is broken).

File: tests/time_change/test [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    filter_setup( $syscall, $key );
+
+    # run the test
+    $result = system("$basedir/test_settime $mode 757094400 123456");
+    ok( $result, 0 );
+
+    # make sure the records had a chance to bubble through to the logs
+    system("auditctl -m syncmarker-$key");
+    for ( my $i = 0 ; $i < 10 ; $i++ ) {
+        if ( system("ausearch -m USER | grep -q syncmarker-$key") eq 0 ) {
```
Comment (pcmoore) [inline tests/time_change/test]: Why are using grep and not `ausearch -k`?

Comment (WOnder93) [reply tests/time_change/test]: Because the marker is a USER record and the key is in its text content, not in the key field.

Comment (pcmoore) [reply tests/time_change/test]: D'oh!  Of course :)  Sorry for the noise.

**Comment by WOnder93** [comment]
The `test_settime( "settimeofday", "settimeofday" );` fails for me as well on Fedora 32... apparently because glibc's settimeofday() now actually calls the clock_settime() syscall under the hood. I'll try replacing the *timeofday() calls with direct syscall(...) calls.

**Comment by WOnder93** [comment]
OK, it seems it worked. @pcmoore please try again with the latest version.

**Comment by pcmoore** [comment]
Merged via ece04ca5517449930cafdddd8c076a258d0e0faf, thanks.

**Pull request closed without merging**